File: core/src/imageops.cpp
// imageops.cpp — ImageEngine 实现，复刻 imageops.py
#include "singan2/imageops.h"

#include <cmath>
#include <stdexcept>

namespace singan2 {

void ImageEngine::set_oneimg(const std::vector<OnebyteImage>& images) {
    for (const auto& im : images) {
        oneimg[im.name] = im.data;
    }
}

void ImageEngine::to_2byte() {
    for (int i = 1; i <= 22; i++) {
        std::string key = "Img" + std::to_string(i);
        auto it = oneimg.find(key);
        if (it != oneimg.end()) {
            std::vector<uint16_t> v(it->second.size());
            for (size_t k = 0; k < it->second.size(); k++) v[k] = it->second[k];
            twoimg[key] = std::move(v);
        }
    }
}

void ImageEngine::compute_intermediate_waves(int red_offset, int grn_offset) {
    if (w_table == nullptr) throw std::runtime_error("ImageEngine.w_table 未设置");
    const auto i1 = oneimg.find("Img1");
    const auto i2 = oneimg.find("Img2");
    const auto i3 = oneimg.find("Img3");
    if (i1 == oneimg.end() || i2 == oneimg.end() || i3 == oneimg.end())
        throw std::runtime_error("compute_intermediate_waves 需要 Img1/Img2/Img3");

    const int N = Y_SIZE * X_SIZE;
    std::vector<int64_t> a1(N), a2(N), a3(N);
    for (int k = 0; k < N; k++) {
        a1[k] = i1->second[k];
        a2[k] = i2->second[k];
        a3[k] = i3->second[k];
    }

    const int w64 = static_cast<int>(w_table->table[64]);
    const int w8 = static_cast<int>(w_table->table[8]);
    auto clamp = [](int64_t v) -> uint8_t {
        if (v < 0) v = 0;
        if (v > 255) v = 255;
        return static_cast<uint8_t>(v);
    };

    std::vector<uint8_t> Img7(N), Img8(N), Img9(N), Img10(N), Img11(N),
        Img12(N), Img13(N), Img14(N), Img15(N);
    for (int k = 0; k < N; k++) {
        Img7[k] = clamp((a1[k] * a1[k]) * w64 >> 16);
        Img8[k] = clamp(a1[k] - a3[k] + red_offset);
        Img9[k] = clamp(a2[k] - a3[k] + grn_offset);
        Img10[k] = clamp(std::abs(a1[k] - a2[k]));
        Img11[k] = clamp(a1[k] - a2[k]);
        Img12[k] = clamp((a1[k] - a2[k]) * (a1[k] - a2[k]) * w8 >> 16);
        Img13[k] = clamp(a2[k] - a1[k]);
        Img14[k] = clamp(a1[k] & a2[k]);
        Img15[k] = clamp(a1[k] | a2[k]);
    }
    oneimg["Img7"] = std::move(Img7);
    oneimg["Img8"] = std::move(Img8);
    oneimg["Img9"] = std::move(Img9);
    oneimg["Img10"] = std::move(Img10);
    oneimg["Img11"] = std::move(Img11);
    oneimg["Img12"] = std::move(Img12);
    oneimg["Img13"] = std::move(Img13);
    oneimg["Img14"] = std::move(Img14);
    oneimg["Img15"] = std::move(Img15);
    to_2byte();
}

const std::vector<uint8_t>& ImageEngine::oneimg_at(int tab) const {
    return oneimg.at("Img" + std::to_string(tab + 1));
}
const std::vector<uint16_t>& ImageEngine::twoimg_at(int tab) const {
    return twoimg.at("Img" + std::to_string(tab + 1));
}
const std::vector<uint16_t>* ImageEngine::twoimg_find(const std::string& name) const {
    auto it = twoimg.find(name);
    return it == twoimg.end() ? nullptr : &it->second;
}
const std::vector<uint8_t>* ImageEngine::oneimg_find(const std::string& name) const {
    auto it = oneimg.find(name);
    return it == oneimg.end() ? nullptr : &it->second;
}
void ImageEngine::to_2byte_orver_write(int tab, const std::vector<uint16_t>& img2byte) {
    twoimg["Img" + std::to_string(tab + 1)] = img2byte;
}

void ImageEngine::gradient(int gtype, int amp) {
    if (w_table == nullptr) throw std::runtime_error("ImageEngine.w_table 未设置");
    const std::vector<uint8_t>& img1byte = oneimg_at(tab_no);
    const std::vector<uint16_t>& img2byte = twoimg_at(tab_no);

    // 边缘填充(常数0)
    std::vector<int64_t> padded((Y_SIZE + 2) * (X_SIZE + 2), 0);
    for (int i = 0; i < Y_SIZE; i++)
        for (int j = 0; j < X_SIZE; j++)
            padded[(i + 1) * (X_SIZE + 2) + (j + 1)] = img1byte[i * X_SIZE + j];

    int cx[3][3] = {0}, cy[3][3] = {0};
    if (gtype == 0) {  // Sobel
        cx[0][0] = -1; cx[0][2] = 1; cx[1][0] = -2; cx[1][2] = 2; cx[2][0] = -1; cx[2][2] = 1;
        cy[0][0] = -1; cy[0][1] = -2; cy[0][2] = -1; cy[2][0] = 1; cy[2][1] = 2; cy[2][2] = 1;
    } else if (gtype == 1) {  // Roberts
        cx[1][1] = 1; cx[2][2] = -1;
        cy[1][2] = 1; cy[2][1] = -1;
    } else {  // 通常
        cx[1][1] = 1; cx[1][2] = -1;
        cy[1][1] = 1; cy[2][1] = -1;
    }

    const int N = Y_SIZE * X_SIZE;
    std::vector<int64_t> zz(N, 0);
    for (int i = 0; i < Y_SIZE; i++) {
        for (int j = 0; j < X_SIZE; j++) {
            int64_t sx = 0, sy = 0;
            for (int ki = 0; ki < 3; ki++)
                for (int kj = 0; kj < 3; kj++) {
                    int64_t v = padded[(i + ki) * (X_SIZE + 2) + (j + kj)];
                    sx += cx[ki][kj] * v;
                    sy += cy[ki][kj] * v;
                }
            int64_t xxyy = sx * sx + sy * sy;
            int64_t val = 0;
            if (gtype == 0 || gtype == 1) {
                int64_t r = rute(static_cast<int>(xxyy), 0, 0x0100, *w_table);
                val = amp * r;
                if (val < 0) val = 0;
                if (val > 255) val = 255;
            } else {
                int64_t r = old_rute(static_cast<int>(xxyy));
                if (r <= 15) r = 0;
                val = (amp * r) & 0xffff;
            }
            zz[i * X_SIZE + j] = val;
        }
    }
    std::vector<uint16_t> result = img2byte;
    for (int i = 1; i < Y_SIZE - 1; i++)
        for (int j = 1; j < X_SIZE - 1; j++)
            result[i * X_SIZE + j] = static_cast<uint16_t>(zz[i * X_SIZE + j]);
    to_2byte_orver_write(tab_no, result);
}

void ImageEngine::niti(int s) {
    const std::vector<uint8_t>& img1byte = oneimg_at(tab_no);
    const int N = Y_SIZE * X_SIZE;
    std::vector<uint16_t> out(N);
    for (int k = 0; k < N; k++) out[k] = (img1byte[k] >= s) ? 0xff : 0;
    to_2byte_orver_write(tab_no, out);
}

void ImageEngine::smooth() {
    if (w_table == nullptr) throw std::runtime_error("ImageEngine.w_table 未设置");
    const std::vector<uint8_t>& img1byte = oneimg_at(tab_no);
    const std::vector<uint16_t>& img2byte = twoimg_at(tab_no);

    std::vector<int64_t> padded((Y_SIZE + 2) * (X_SIZE + 2), 0);
    for (int i = 0; i < Y_SIZE; i++)
        for (int j = 0; j < X_SIZE; j++)
            padded[(i + 1) * (X_SIZE + 2) + (j + 1)] = img1byte[i * X_SIZE + j];

    const int w9 = static_cast<int>(w_table->table[9]);
    const int N = Y_SIZE * X_SIZE;
    std::vector<int64_t> out(N, 0);
    for (int i = 0; i < Y_SIZE; i++) {
        for (int j = 0; j < X_SIZE; j++) {
            int64_t buf = padded[i * (X_SIZE + 2) + j] + padded[i * (X_SIZE + 2) + j + 1] + padded[i * (X_SIZE + 2) + j + 2]
                        + padded[(i + 1) * (X_SIZE + 2) + j] + padded[(i + 1) * (X_SIZE + 2) + j + 1] + padded[(i + 1) * (X_SIZE + 2) + j + 2]
                        + padded[(i + 2) * (X_SIZE + 2) + j] + padded[(i + 2) * (X_SIZE + 2) + j + 1] + padded[(i + 2) * (X_SIZE + 2) + j + 2];
            int64_t v = (buf * w9) >> 16;
            v += img1byte[i * X_SIZE + j];
            if (v < 0) v = 0;
            if (v > 255) v = 255;
            out[i * X_SIZE + j] = v;
        }
    }
    std::vector<uint16_t> result = img2byte;
    for (int i = 1; i < Y_SIZE - 1; i++)
        for (int j = 1; j < X_SIZE - 1; j++)
            result[i * X_SIZE + j] = static_cast<uint16_t>(out[i * X_SIZE + j]);
    to_2byte_orver_write(tab_no, result);
}

int ImageEngine::rute(int u_bunsan, int left, int right, const WTable& wt) {
    int center = 0x80;
    int w2 = static_cast<int>(wt.table[2]);
    for (int it = 0; it < 7; it++) {
        if (center * center < u_bunsan)
            left = center;
        else
            right = center;
        center = ((left + right) * w2) >> 16;
    }
    left = center;
    center = ((left + right) * w2) >> 16;
    return center;
}

int ImageEngine::old_rute(int u_bunsan) {
    int i = 0;
    while (true) {
        if (i * i > u_bunsan) return i - 1;
        i++;
    }
}

}  // namespace singan2

File: poc/algo/imageops.py
# -*- coding: utf-8 -*-
"""
imageops.py — SINGAN2 图像处理基础函数（Python 复刻）

复刻来源（按文件）：
  - GRADIENT.CPP  :: Gradient(type, amp)        Sobel/Roberts/通常 梯度
  - NITI.CPP      :: NITI(s)                    单纯二值化
  - smooth_median.cpp :: smooth()              移动平均法
  - To2byte.cpp   :: To2byte()                 1字节->2字节拷贝
  - To2byte_orver_write.cpp :: To2byte_orver_write() 处理结果写回 global_twoimg
  - MainRun.cpp   :: Inimg()                    按 TabNo 选择原始图像
  - bvmath.cpp    :: Rute() / Old_Rute()        DSP 平方根

设计：
  - 全部用 numpy 向量化，结果与 C++ 逐像素计算一致
  - 图像以二维数组 (Y_SIZE, X_SIZE) 或一维数组表示
"""
import math
import numpy as np
import logging
import traceback

# ----------------------------------------------------------------------------
# 常量（与 MAIN.H / 各算法文件一致）
# ----------------------------------------------------------------------------
Y_SIZE = 88
X_SIZE = 186
ONESIZE = Y_SIZE * X_SIZE

# w_Table 大小（GBV_DIV_H.bin 大小 / 2）
W_TABLE_SIZE = 32768 // 2


def get_logger():
    return logging.getLogger("imageops")


class ImageEngine:
    """图像处理引擎：持有 global_oneimg / global_twoimg / w_Table 等全局状态。

    对应 C++ 全局量：global_oneimg, global_twoimg, global_TabNo, w_Table。
    """

    def __init__(self, w_table=None):
        # global_oneimg: 1字节图像 (Y_SIZE, X_SIZE)，名 Img1..Img15
        self.oneimg = {}
        # global_twoimg: 2字节图像 (Y_SIZE, X_SIZE)
        self.twoimg = {}
        # 当前处理波段号（对应 global_TabNo）
        self.tab_no = 0
        # 除法表 w_Table（长度 16384，w_Table[n] 约等于 65536/n）
        self.w_table = w_table if w_table is not None else np.zeros(W_TABLE_SIZE, dtype=np.uint16)

    # ------------------------------------------------------------------//
    # 数据准备
    # ------------------------------------------------------------------//
    def set_oneimg(self, img_dict):
        """从波段图像字典填充 global_oneimg（Img1..Img22）。

        参数:
            img_dict: {波段名: np.ndarray(ONESIZE,)} 来自 mariner_reader.build_onebyte_images
        """
        try:
            for name, buf in img_dict.items():
                arr = np.frombuffer(bytes(buf), dtype=np.uint8).reshape(Y_SIZE, X_SIZE)
                self.oneimg[name] = arr
            logger = get_logger()
            logger.debug("oneimg 波段: %s", sorted(self.oneimg.keys()))
        except Exception as exc:
            get_logger().error("set_oneimg 异常: %s\n%s", exc, traceback.format_exc())
            raise

    def to_2byte(self):
        """To2byte(): 把 global_oneimg 的 Img1..Img22 拷贝到 global_twoimg。"""
        try:
            for i in range(1, 23):
                key = "Img%d" % i
                if key in self.oneimg:
                    self.twoimg[key] = self.oneimg[key].astype(np.uint16)
        except Exception as exc:
            get_logger().error("to_2byte 异常: %s\n%s", exc, traceback.format_exc())
            raise

    def compute_intermediate_waves(self, red_offset=128, grn_offset=128):
        """ReadImgDataNew 中间波段计算（MainRun.cpp:908-1021 对应部分）。

        计算 Img7..Img15（写入 oneimg），并把 Img1..22 全部拷贝到 twoimg。
           Img7  = Img1^2 / 64            (clamp 0..255)
           Img8  = Img1 - Img3 + red_offset   (clamp 0..255)
           Img9  = Img2 - Img3 + grn_offset   (clamp 0..255)
           Img10 = abs(Img1 - Img2)           (clamp 0..255)
           Img11 = Img1 - Img2                (clamp 0..255)
           Img12 = (Img1-Img2)^2 / 8          (clamp 0..255)
           Img13 = Img2 - Img1                (clamp 0..255)
           Img14 = Img1 & Img2
           Img15 = Img1 | Img2
        """
        try:
            i1 = self.oneimg["Img1"].astype(np.int64)
            i2 = self.oneimg["Img2"].astype(np.int64)
            i3 = self.oneimg["Img3"].astype(np.int64)

            def clamp(dat):
                return np.clip(dat, 0, 255).astype(np.uint8)

            w64 = int(self.w_table[64]) if len(self.w_table) > 64 else 0
            w8 = int(self.w_table[8]) if len(self.w_table) > 8 else 0

            self.oneimg["Img7"] = clamp((i1 * i1) * w64 >> 16)
            self.oneimg["Img8"] = clamp(i1 - i3 + red_offset)
            self.oneimg["Img9"] = clamp(i2 - i3 + grn_offset)
            self.oneimg["Img10"] = clamp(np.abs(i1 - i2))
            self.oneimg["Img11"] = clamp(i1 - i2)
            self.oneimg["Img12"] = clamp((i1 - i2) * (i1 - i2) * w8 >> 16)
            self.oneimg["Img13"] = clamp(i2 - i1)
            self.oneimg["Img14"] = clamp(np.bitwise_and(self.oneimg["Img1"].astype(np.int64),
                                                        self.oneimg["Img2"].astype(np.int64)))
            self.oneimg["Img15"] = clamp(np.bitwise_or(self.oneimg["Img1"].astype(np.int64),
                                                       self.oneimg["Img2"].astype(np.int64)))
            self.to_2byte()
            get_logger().debug("compute_intermediate_waves 完成: oneimg=%s",
                               sorted(self.oneimg.keys()))
        except Exception as exc:
            get_logger().error("compute_intermediate_waves 异常: %s\n%s", exc, traceback.format_exc())
            raise

    def inimg(self):
        """Inimg(): 按 global_TabNo 选择原始图像，返回 (img1byte, img2byte)。

        img1byte: 1字节图像 (Y_SIZE, X_SIZE)
        img2byte: 2字节图像 (Y_SIZE, X_SIZE)
        """
        try:
            tab = self.tab_no
            key1 = "Img%d" % (tab + 1)
            key2 = "Img%d" % (tab + 1)
            if key1 not in self.oneimg:
                raise ValueError("oneimg 缺少 %s" % key1)
            img1byte = self.oneimg[key1].copy()
            img2byte = self.twoimg.get(key2, np.zeros((Y_SIZE, X_SIZE), dtype=np.uint16)).copy()
            return img1byte, img2byte
        except Exception as exc:
            get_logger().error("inimg 异常: %s\n%s", exc, traceback.format_exc())
            raise

    def to_2byte_orver_write(self, img2byte):
        """To2byte_orver_write(): 处理结果写回 global_twoimg.Img{tab_no+1}。"""
        try:
            tab = self.tab_no
            key = "Img%d" % (tab + 1)
            self.twoimg[key] = np.asarray(img2byte, dtype=np.uint16).reshape(Y_SIZE, X_SIZE)
        except Exception as exc:
            get_logger().error("to_2byte_orver_write 异常: %s\n%s", exc, traceback.format_exc())
            raise

    # ------------------------------------------------------------------//
    # 梯度
    # ------------------------------------------------------------------//
    @staticmethod
    def _gradient_kernels(gtype):
        """返回 (cx, cy) 3x3 卷积核。gtype: 0=Sobel 1=Roberts 2=通常"""
        if gtype == 0:
            cx = np.array([[-1, 0, 1], [-2, 0, 2], [-1, 0, 1]], dtype=np.int64)
            cy = np.array([[-1, -2, -1], [0, 0, 0], [1, 2, 1]], dtype=np.int64)
        elif gtype == 1:
            cx = np.array([[0, 0, 0], [0, 1, 0], [0, 0, -1]], dtype=np.int64)
            cy = np.array([[0, 0, 0], [0, 0, 1], [0, -1, 0]], dtype=np.int64)
        else:  # 2: 通常
            cx = np.array([[0, 0, 0], [0, 1, -1], [0, 0, 0]], dtype=np.int64)
            cy = np.array([[0, 0, 0], [0, 1, 0], [0, -1, 0]], dtype=np.int64)
        return cx, cy

    def gradient(self, gtype, amp):
        """Gradient(type, amp): 梯度处理，结果写回 twoimg.Img{tab_no+1}。

        对应 GRADIENT.CPP Gradient()。逐像素:
          xx = sum(cx[k]*d[k])  yy = sum(cy[k]*d[k])
          case 0/1: xxyy = Rute(xx^2+yy^2, 0, 0x100); zz=amp*xxyy; clamp [0,255]
          case 2:   xxyy = Old_Rute(xx^2+yy^2); if<=15 ->0; zz=amp*xxyy; (USHORT截断)
        仅处理内部区域 i∈[1,Y-2], j∈[1,X-2]；边界像素保持 twoimg 原始值（Inimg 拷贝）。
        """
        try:
            img1byte, img2byte = self.inimg()
            cx, cy = self._gradient_kernels(gtype)
            # 扩展图像边缘（C++ 从 i=1..Y-2, j=1..X-2）
            padded = np.pad(img1byte.astype(np.int64), 1, mode="constant", constant_values=0)
            # 计算 xx, yy
            xx = (cx[0, 0] * padded[0:-2, 0:-2] + cx[0, 1] * padded[0:-2, 1:-1] + cx[0, 2] * padded[0:-2, 2:]
                  + cx[1, 0] * padded[1:-1, 0:-2] + cx[1, 1] * padded[1:-1, 1:-1] + cx[1, 2] * padded[1:-1, 2:]
                  + cx[2, 0] * padded[2:, 0:-2] + cx[2, 1] * padded[2:, 1:-1] + cx[2, 2] * padded[2:, 2:])
            yy = (cy[0, 0] * padded[0:-2, 0:-2] + cy[0, 1] * padded[0:-2, 1:-1] + cy[0, 2] * padded[0:-2, 2:]
                  + cy[1, 0] * padded[1:-1, 0:-2] + cy[1, 1] * padded[1:-1, 1:-1] + cy[1, 2] * padded[1:-1, 2:]
                  + cy[2, 0] * padded[2:, 0:-2] + cy[2, 1] * padded[2:, 1:-1] + cy[2, 2] * padded[2:, 2:])
            xxyy = xx * xx + yy * yy

            if gtype in (0, 1):
                # Rute 二分迭代平方根（Rute(xxyy, 0, 0x100)）
                out = np.empty_like(xxyy, dtype=np.int64)
                for idx in range(xxyy.size):
                    out.flat[idx] = self._rute(int(xxyy.flat[idx]), 0, 0x0100, self.w_table)
                # 利得: zz = amp * xxyy; clamp [0,255]
                zz = amp * out
                zz = np.clip(zz, 0, 255)
            else:
                # Old_Rute 线性平方根
                out = np.empty_like(xxyy, dtype=np.int64)
                for idx in range(xxyy.size):
                    v = int(xxyy.flat[idx])
                    r = self._old_rute(v)
                    out.flat[idx] = r if r > 15 else 0
                # 利得: zz = amp * xxyy; USHORT 截断(溢出取低16位)
                zz = (amp * out) & 0xffff
            # 只写内部区域，边界保持 twoimg 原始值
            result = np.asarray(img2byte, dtype=np.int64).copy()
            result[1:Y_SIZE - 1, 1:X_SIZE - 1] = zz[1:Y_SIZE - 1, 1:X_SIZE - 1]
            self.to_2byte_orver_write(result.astype(np.uint16))
        except Exception as exc:
            get_logger().error("gradient 异常: %s\n%s", exc, traceback.format_exc())
            raise

    @staticmethod
    def _rute(u_bunsan, left_threashold, right_threashold, w_table):
        """Rute(): 二分迭代平方根（bvmath.cpp Rute）。

        7 次迭代：if center^2 < uBunsan: left=center else right=center
        然后 center = (left+right)*w_Table[2]>>16。
        """
        center = 0x80
        w2 = int(w_table[2]) if w_table is not None else 32768
        for _ in range(7):
            if center * center < u_bunsan:
                left_threashold = center
                center = ((left_threashold + right_threashold) * w2) >> 16
            else:
                right_threashold = center
                center = ((left_threashold + right_threashold) * w2) >> 16
        left_threashold = center
        center = ((left_threashold + right_threashold) * w2) >> 16
        return center

    @staticmethod
    def _old_rute(u_bunsan):
        """Old_Rute(): 线性平方根（bvmath.cpp Old_Rute）。"""
        i = 0
        while True:
            if i * i > u_bunsan:
                return i - 1
            i += 1

    # ------------------------------------------------------------------//
    # 二值化
    # ------------------------------------------------------------------//
    def niti(self, s):
        """NITI(s): 单纯二值化，>=s 置 0xff，否则 0，写回 twoimg.Img{tab_no+1}。"""
        try:
            img1byte, _ = self.inimg()
            out = np.where(img1byte >= s, 0xff, 0).astype(np.uint16)
            self.to_2byte_orver_write(out)
        except Exception as exc:
            get_logger().error("niti 异常: %s\n%s", exc, traceback.format_exc())
            raise

    # ------------------------------------------------------------------//
    # 移动平均（smooth）
    # ------------------------------------------------------------------//
    def smooth(self):
        """smooth(): 移动平均（smooth_median.cpp:61-110）。

        仅处理内部区域 i∈[1,Y-2], j∈[1,X-2]：
          buf = (3x3 和) * w_Table[9] >> 16 + 中心像素；clamp 255
        边界像素保持 twoimg 原始值。
        """
        try:
            img1byte, img2byte = self.inimg()
            padded = np.pad(img1byte.astype(np.int64), 1, mode="constant", constant_values=0)
            buf = (padded[0:-2, 0:-2] + padded[0:-2, 1:-1] + padded[0:-2, 2:]
                   + padded[1:-1, 0:-2] + padded[1:-1, 1:-1] + padded[1:-1, 2:]
                   + padded[2:, 0:-2] + padded[2:, 1:-1] + padded[2:, 2:])
            # buf * w_Table[9] >> 16（w_Table[9]≈7282），再加中心像素，clamp 255
            out = (buf * int(self.w_table[9])) >> 16
            out += img1byte.astype(np.int64)
            out = np.clip(out, 0, 255)
            result = np.asarray(img2byte, dtype=np.int64).copy()
            result[1:Y_SIZE - 1, 1:X_SIZE - 1] = out[1:Y_SIZE - 1, 1:X_SIZE - 1]
            self.to_2byte_orver_write(result.astype(np.uint16))
        except Exception as exc:
            get_logger().error("smooth 异常: %s\n%s", exc, traceback.format_exc())
            raise
